src/states/main_game.rs

```diff
@@ -213,6 +226,33 @@ impl SimpleState for MainGameState {
 
     fn update(&mut self, data: &mut StateData<'_, GameData<'_, '_>>) -> SimpleTrans {
         self.dispatcher.dispatch(&mut data.world.res);
+
+        let input_event_channel = data
+            .world
+            .read_resource::<EventChannel<InputEvent<String>>>();
+        for event in input_event_channel.read(self.input_event_reader_id.as_mut().unwrap()) {
+            match event {
+                InputEvent::ActionPressed(action_name) => {
+                    if action_name == "Pause" {
+                        return Trans::Push(Box::new(PausedState::default()));
+                    }
+                }
+                _ => (),
+            }
+        }
         Trans::None
     }
+
+    fn on_resume(&mut self, data: StateData<'_, GameData<'_, '_>>) {
+        {
```

Comment:
Could you add comments on both `on_resume` functions as to why you are re-registering the reader?